rocAL/source/meta_data/webdataset_meta_data_reader.cpp

```diff
@@ -0,0 +1,435 @@
+                "directory at " +
+                _folder_path);
+        _full_path = _folder_path;
+        while ((_entity = readdir(_sub_dir)) != nullptr) {
+            std::string entry_name(_entity->d_name);
+            if (strcmp(_entity->d_name, ".") == 0 || strcmp(_entity->d_name, "..") == 0)
+                continue;
+            _index_name_list.push_back(entry_name);
+        }
+        std::sort(_index_name_list.begin(), _index_name_list.end());
+        if ((_sub_dir = opendir(_path.c_str())) == nullptr)
+            THROW(
+                "WebDatasetSourceReader :: ERROR: Failed opening the "
+                "directory at " +
+                _path);
+        _full_path = _path;
+        while ((_entity = readdir(_sub_dir)) != nullptr) {
+            std::string entry_name(_entity->d_name);
+            if (strcmp(_entity->d_name, ".") == 0 || strcmp(_entity->d_name, "..") == 0)
+                continue;
+            entry_name_list.push_back(entry_name);
+        }
+        std::sort(entry_name_list.begin(), entry_name_list.end());
+        _wds_shards.reserve(entry_name_list.size());
+        for (auto &path : entry_name_list)
+            _wds_shards.emplace_back(FileIOStream::open(_path + path));
+    }
+    closedir(_sub_dir);
+
+    std::vector<SampleDescription> unfiltered_samples;
+    std::vector<ComponentDescription> unfiltered_components;
+
+    for (unsigned wds_shard_index = 0; wds_shard_index < entry_name_list.size();
+         ++wds_shard_index) {
+        unfiltered_samples.resize(0);
+        unfiltered_components.resize(0);
+        if (_index_paths.size() == 0)
+            parse_tar_files(unfiltered_samples, unfiltered_components,
+                            _wds_shards[wds_shard_index]);
+        else
+            parse_index_files(unfiltered_samples, unfiltered_components,
+                              _folder_path + _index_name_list[wds_shard_index]);
+
+        // After parsing add the contents to the map
+        for (auto &sample : unfiltered_samples) {
+            if (_missing_component_behaviour == MissingComponentsBehaviour::EMPTY) {  // empty_outputs
+                AsciiValues ascii_values;
+                ascii_values.resize(_ext_map.size());
+                std::string last_file_name;
+                for (auto &component : sample.components) {
+                    if (!isJPEG(component.ext)) {  // Add more components as we encounter
+                        _wds_shards[wds_shard_index]->set_read_position(component.offset);
+                        std::vector<uint8_t> cls_data(component.size);
+                        _wds_shards[wds_shard_index]->read_into_buffer(cls_data.data(), component.size);
+                        AsciiComponent ascii_component = {};
+                        for (size_t i = 0; i < cls_data.size(); ++i)
+                            ascii_component.push_back(static_cast<uint8_t>(cls_data[i]));
```

Comment:
this is very inefficient way. Please use uint8_t * instead of AsciiComponent and use memcpy for all cls_data.size() as one instruction.
I see all new code is overly using std::vector types for even simplest data-type. remember you are calling a function of a class every time when you push a byte into the vector!

Reply:
Changed to using shared ptr